dali/operators/reader/loader/tar_utils.cc

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/tar_utils.h"
+#include <libtar.h>
+#include <algorithm>
+#include <cstdarg>
+#include <cstdlib>
+#include <string>
+#include <tuple>
+#include <utility>
+#include "dali/core/error_handling.h"
+
+namespace dali {
+namespace detail {
+
+constexpr uint64 kBlockSize = 512;
+constexpr uint64 kEmptyEofBlocks = 2;
+
+std::mutex instances_mutex;
+vector<TarArchive*> instances;
+
+int Register(TarArchive* archive) {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  int instance_handle = -1;
+  for (auto& instances_entry : instances) {
+    if (instances_entry != nullptr) {
+      continue;
+    }
+    instances_entry = archive;
+    instance_handle = &instances_entry - instances.data();
+  }
+  if (instance_handle == -1) {
+    instance_handle = instances.size();
+    instances.push_back(archive);
```

Comment:
This makes accessing the array for reading non thread-safe. The vector should be allocated once and "big enough" or you could even use a plain array - growing while it's not empty it will invalidate the pointer and a race condition might occur between `push_back` and unguarded `operator[]`.